Python/compile.c

```diff
@@ -253,25 +274,58 @@ is_jump(struct instr *i)
     return IS_JUMP_OPCODE(i->i_opcode);
 }
 
+static oparg_type
+opcode_to_oparg_type(int opcode) {
+    if (HAS_TARGET(opcode)) {
+        return JUMP_TARGET_ARG;
+    }
+    else if (HAS_CONST(opcode)) {
+        return CONST_ARG;
+    }
+    else if (HAS_ARG(opcode)) {
+        return EXPLICIT_ARG;
+    }
+    else {
+        return UNUSED_ARG_TYPE;
+    }
+    Py_UNREACHABLE();
+}
+
+#ifdef Py_DEBUG
+static bool
+arg_types_are_compatible(struct instr *instr) {
+    int opcode = instr->i_opcode;
+    oparg_t oparg = instr->i_oparg;
+    oparg_type expected = opcode_to_oparg_type(opcode);
+    if (oparg.type != expected) {
+        fprintf(stderr,
+                "opcode = %d oparg.type = %d expected %d\n",
+                opcode, oparg.type, expected);
+        return false;
+    }
+    return true;
+}
+#endif
+
 static int
-instr_size(struct instr *instruction)
+instr_size(struct instr *instr)
 {
-    int opcode = instruction->i_opcode;
+    int opcode = instr->i_opcode;
     assert(!IS_PSEUDO_OPCODE(opcode));
-    int oparg = instruction->i_oparg;
-    assert(HAS_ARG(opcode) || oparg == 0);
+    assert(arg_types_are_compatible(instr));
+    int oparg = OPARG_VALUE(instr->i_oparg);
```

Comment:
perhaps `assert(arg_types_are_compatible(instr));` may as well just be built-in to the `OPARG_VALUE` macro, ensuring you can never get an oparg value of the wrong type (in a debug build)?